include/SFML/System/Angle.hpp

```diff
@@ -0,0 +1,473 @@
+// In no event will the authors be held liable for any damages arising from the use of this software.
+//
+// Permission is granted to anyone to use this software for any purpose,
+// including commercial applications, and to alter it and redistribute it freely,
+// subject to the following restrictions:
+//
+// 1. The origin of this software must not be misrepresented;
+//    you must not claim that you wrote the original software.
+//    If you use this software in a product, an acknowledgment
+//    in the product documentation would be appreciated but is not required.
+//
+// 2. Altered source versions must be plainly marked as such,
+//    and must not be misrepresented as being the original software.
+//
+// 3. This notice may not be removed or altered from any source distribution.
+//
+////////////////////////////////////////////////////////////
+
+#ifndef SFML_ANGLE_HPP
+#define SFML_ANGLE_HPP
+
+////////////////////////////////////////////////////////////
+// Headers
+////////////////////////////////////////////////////////////
+#include <SFML/System/Export.hpp>
+#include <cassert>
+
+
+namespace sf
+{
+////////////////////////////////////////////////////////////
+/// \brief Represents an angle value.
+///
+////////////////////////////////////////////////////////////
+class Angle
+{
+public:
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Default constructor
+    ///
+    /// Sets the angle value to zero.
+    ///
+    ////////////////////////////////////////////////////////////
+    constexpr Angle();
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Return the angle's value in degrees
+    ///
+    /// \return Angle in degrees
+    ///
+    /// \see asRadians
+    ///
+    ////////////////////////////////////////////////////////////
+    [[nodiscard]] constexpr float asDegrees() const;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Return the angle's value in radians
+    ///
+    /// \return Angle in radians
+    ///
+    /// \see asDegrees
+    ///
+    ////////////////////////////////////////////////////////////
+    [[nodiscard]] constexpr float asRadians() const;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Return the angle wrapped to the range [-180°, 180°)
+    ///
+    /// \return Angle wrapped to [-180°, 180°)
+    ///
+    /// \see wrappedTo360
+    ///
+    ////////////////////////////////////////////////////////////
+    [[nodiscard]] constexpr Angle wrappedTo180() const;
+
+    ////////////////////////////////////////////////////////////
+    /// \brief Return the angle wrapped to the range [0°, 360°)
+    ///
+    /// \return Angle wrapped to [0°, 360°)
+    ///
+    /// \see wrappedTo180
+    ///
+    ////////////////////////////////////////////////////////////
+    [[nodiscard]] constexpr Angle wrappedTo360() const;
```

Comment:
Actually just realized that the naming is also not very good, when you consider that Angle is now a unit agnostic class, so assuming 360 to mean degree and encode it on a function call, kind of works against this (why no `wrappedToTau`? 😉).
As such I'd recommend picking someone other naming like `wrappedToFullRotation` or similar.

Reply:
Yeah, that was also one of my concerns. The thing is, both are full rotations, just with different start/end points. I was also thinking about `wrappedToTau()` and `wrappedToPi()` (where tau = 2*pi), but this is even less known and suffers from the same issue, just in radians.

Another idea I had was to use "signed/unsigned". E.g. `wrapSigned()` and `wrapUnsigned()`.

From a readability point of view, "un" is quite a small difference (in more complex expressions), so having uppercase "S" and "U" helps quickly differentiate the two.

Reply:
But I'd have no clue what `wrapSigned()` means from an API perspective and the -180, 180 range would contain both signed and unsigned values.

Another idea I had was focusing more on the unit circle, so wrap FullCircle or HalfCircle, but that might be rather confusing as well.

I guess 180 vs 360 is the most intuitive way, but we'd then have to state that this is of course not limited to just degrees (which ofc doesn't even make sense since `sf::Angle` is unit agnostic...).

Reply:
> But I'd have no clue what wrapSigned() means from an API perspective 

I think in general this concept is not well-known enough that we could make it 100% obvious in 1-2 words. There is not even a name for either of the two conventions. So maybe we should focus our efforts more on recognizability (once you know the function) than self-explanatory-ness. Reading the documentation once and understanding it from then on is acceptable in my opinion.

> and the -180, 180 range would contain both signed and unsigned values.

That's exactly what "signed" means, it has a sign 🙂 you probably meant "positive and negative".

The reason why "signed/unsigned" are quite elegant, is that their ranges behave _precisely_ like those of signed/unsigned integers (just with float):

|| signed | unsigned |
|---|---|---|
| char | [-128, 128( | [0, 256( |
| Angle| [-180°, 180°( | [0°, 360°( |

Reply:
Maybe that discussion we're having about naming rationale and analogy to signed/unsigned could indeed fit the detailed section of the doc (between `\brief` and `\return`).

```cpp
    ////////////////////////////////////////////////////////////
    /// \brief Wrap to a range, so that -180° <= angle < 180°
    /// 
    /// Similar to a modulo operation, this returns a copy of the angle
    /// constrained to the range [-180°, 180°) == [-Pi, Pi).
    /// The resulting angle represents a rotation which is equivalent to *this.
    ///
    /// The name "signed" originates from the similarity to signed integers:
    /// signed char for example has range [-128, 128), just with discrete steps.
    /// In contrast, unsigned char has range [0, 256).
    ///
    /// \return Signed angle, wrapped to [-180°, 180°)
    /// \see wrapUnsigned
    ///
    ////////////////////////////////////////////////////////////
    [[nodiscard]] constexpr Angle wrapSigned() const;


    ////////////////////////////////////////////////////////////
    /// \brief Wrap to a range, so that 0° <= angle < 360°
    /// 
    /// Similar to a modulo operation, this returns a copy of the angle
    /// constrained to the range [0°, 360°) == [0, Tau) == [0, 2*Pi).
    /// The resulting angle represents a rotation which is equivalent to *this.
    ///
    /// The name "unsigned" originates from the similarity to unsigned integers:
    /// unsigned char for example has range [0, 256), just with discrete steps.
    /// In contrast, signed char has range [-128, 128).
    ///
    /// \return Unsigned angle, wrapped to [0°, 360°)
    /// \see wrapSigned
    ///
    ////////////////////////////////////////////////////////////
    [[nodiscard]] constexpr Angle wrapUnsigned() const;
```

Reply:
> The reason why "signed/unsigned" are quite elegant, is that their ranges behave precisely like those of signed/unsigned integers, just with floating point types instead of integers:

I think this is the best argument for a specific name I've heard so far. I like how it borrows terms already familiar to C++ programmers and accurately replicated the behavior of signed and unsigned integers. Quite elegant :)